File: exercise/day17/src/main/java/games/FizzBuzz.java
package games;

import io.vavr.collection.LinkedHashMap;
import io.vavr.collection.Map;
import io.vavr.control.Option;

import static io.vavr.control.Option.none;
import static io.vavr.control.Option.some;

public class FizzBuzz {
    public static final int MIN = 1;
    public static final int MAX = 100;
    private static final Map<Integer, String> mapping = LinkedHashMap.of(
            15, "FizzBuzz",
            3, "Fizz",
            5, "Buzz"
    );

    public static Option<String> convert(int input) {
        return isOutOfRange(input)
                ? none()
                : some(convertSafely(input));
    }

    private static String convertSafely(Integer input) {
        return mapping
                .find(p -> is(p._1, input))
                .map(p -> p._2)
                .getOrElse(input.toString());
    }

    private static boolean is(Integer divisor, Integer input) {
        return input % divisor == 0;
    }

    private static boolean isOutOfRange(Integer input) {
        return input < MIN || input > MAX;
    }
}

File: solution/ts/day17/src/fizzbuzz.ts
import {none, Option, some} from "fp-ts/Option";

export const min = 1;
export const max = 100;

let mapping: Map<number, string> = new Map([
    [15, 'FizzBuzz'],
    [3, 'Fizz'],
    [5, 'Buzz'],
]);

export const fizzbuzz = (input: number): Option<string> =>
    isOutOfRange(input)
        ? none
        : some(convertSafely(input));

function convertSafely(input: number): string {
    for (const [divisor, value] of mapping) {
        if (is(divisor, input)) {
            return value;
        }
    }
    return input.toString();
}

const is = (divisor: number, input: number): boolean => input % divisor === 0;
const isOutOfRange = (input: number): boolean => input < min || input > max;